File: challenges/2026-09-13-merged-step-count-timeline/solution.cs
using System;
using System.Collections.Generic;

public class ListNode
{
    public int Timestamp;
    public int Steps;
    public ListNode Next;
    public ListNode(int timestamp = 0, int steps = 0, ListNode next = null)
    {
        Timestamp = timestamp;
        Steps = steps;
        Next = next;
    }
}

public class Solution
{
    public static ListNode BuildList(List<(int Timestamp, int Steps)> values)
    {
        var dummy = new ListNode();
        var cur = dummy;
        foreach (var (timestamp, steps) in values)
        {
            cur.Next = new ListNode(timestamp, steps);
            cur = cur.Next;
        }
        return dummy.Next;
    }

    public static List<(int Timestamp, int Steps)> ToList(ListNode head)
    {
        var result = new List<(int, int)>();
        var node = head;
        while (node != null)
        {
            result.Add((node.Timestamp, node.Steps));
            node = node.Next;
        }
        return result;
    }

    public static ListNode MergeStepTimelines(ListNode headA, ListNode headB)
    {
        // TODO: implement
        return null;
    }

    private static string FormatPairs(List<(int, int)> pairs)
    {
        var parts = new List<string>();
        foreach (var (a, b) in pairs)
        {
            parts.Add($"({a}, {b})");
        }
        return "[" + string.Join(", ", parts) + "]";
    }

    public static void Main(string[] args)
    {
        var tests = new List<(string Name, List<(int, int)> A, List<(int, int)> B, List<(int, int)> Expected)>
        {
            ("example 1",
                new List<(int, int)> { (0, 0), (30, 1200), (60, 2500) },
                new List<(int, int)> { (15, 600), (45, 1900) },
                new List<(int, int)> { (0, 0), (15, 600), (30, 1200), (45, 1900), (60, 2500) }),
            ("example 2: tie at timestamp 10",
                new List<(int, int)> { (10, 500), (20, 1000) },
                new List<(int, int)> { (10, 480), (25, 1100) },
                new List<(int, int)> { (10, 500), (10, 480), (20, 1000), (25, 1100) }),
            ("example 3: A empty",
                new List<(int, int)>(),
                new List<(int, int)> { (5, 200), (10, 450) },
                new List<(int, int)> { (5, 200), (10, 450) }),
            ("edge: both empty",
                new List<(int, int)>(),
                new List<(int, int)>(),
                new List<(int, int)>()),
            ("edge: B empty",
                new List<(int, int)> { (1, 10), (2, 20) },
                new List<(int, int)>(),
                new List<(int, int)> { (1, 10), (2, 20) }),
        };

        foreach (var (name, a, b, expected) in tests)
        {
            var headA = BuildList(a);
            var headB = BuildList(b);
            var merged = MergeStepTimelines(headA, headB);
            var actual = ToList(merged);

            bool ok = actual.Count == expected.Count;
            if (ok)
            {
                for (int i = 0; i < actual.Count; i++)
                {
                    if (actual[i] != expected[i]) { ok = false; break; }
                }
            }
            string status = ok ? "PASS" : "FAIL";
            Console.WriteLine($"{status} {name}: expected={FormatPairs(expected)} actual={FormatPairs(actual)}");
        }
    }
}

File: challenges/2026-09-13-merged-step-count-timeline/solution.go
package main

import "fmt"

type ListNode struct {
	Timestamp int
	Steps     int
	Next      *ListNode
}

func buildList(values [][2]int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range values {
		cur.Next = &ListNode{Timestamp: v[0], Steps: v[1]}
		cur = cur.Next
	}
	return dummy.Next
}

func toList(head *ListNode) [][2]int {
	result := [][2]int{}
	for node := head; node != nil; node = node.Next {
		result = append(result, [2]int{node.Timestamp, node.Steps})
	}
	return result
}

func mergeStepTimelines(headA, headB *ListNode) *ListNode {
	// TODO: implement
	return nil
}

func equalPairs(a, b [][2]int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

type testCase struct {
	name     string
	a        [][2]int
	b        [][2]int
	expected [][2]int
}

func main() {
	tests := []testCase{
		{"example 1",
			[][2]int{{0, 0}, {30, 1200}, {60, 2500}},
			[][2]int{{15, 600}, {45, 1900}},
			[][2]int{{0, 0}, {15, 600}, {30, 1200}, {45, 1900}, {60, 2500}}},
		{"example 2: tie at timestamp 10",
			[][2]int{{10, 500}, {20, 1000}},
			[][2]int{{10, 480}, {25, 1100}},
			[][2]int{{10, 500}, {10, 480}, {20, 1000}, {25, 1100}}},
		{"example 3: A empty",
			[][2]int{},
			[][2]int{{5, 200}, {10, 450}},
			[][2]int{{5, 200}, {10, 450}}},
		{"edge: both empty",
			[][2]int{},
			[][2]int{},
			[][2]int{}},
		{"edge: B empty",
			[][2]int{{1, 10}, {2, 20}},
			[][2]int{},
			[][2]int{{1, 10}, {2, 20}}},
	}

	for _, tc := range tests {
		headA := buildList(tc.a)
		headB := buildList(tc.b)
		merged := mergeStepTimelines(headA, headB)
		actual := toList(merged)
		status := "FAIL"
		if equalPairs(actual, tc.expected) {
			status = "PASS"
		}
		fmt.Printf("%s %s: expected=%v actual=%v\n", status, tc.name, tc.expected, actual)
	}
}
